File: 2020-hack/feb14/Library.java
package feb14;
import java.util.*;
public class Library {
  private static int[]     books    = new int[10001];
  private static int       MAX_BOOK = 1001;
  private static short[][] aux      = new short[40000][];
  
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int tc = sc.nextInt();
    while (tc-- > 0) {
      int n = sc.nextInt();
      for (int i = 0; i < n; i++) {
        books[i] = sc.nextInt();
      }
      buildRMQ(1, n, 1);
      int q = sc.nextInt();
      for (int i = 0; i < q; i++) {
        switch (sc.nextInt()) {
          case 0 : {
            int from = sc.nextInt();
            int to = sc.nextInt();
            short[] result = new short[MAX_BOOK];
            queryRMQ(1, n, 1, from, to, result);
            int k = sc.nextInt();
            int rank = 0;
            for (int j = 0; j < MAX_BOOK; j++) {
              rank += result[j];
              if (k <= rank) {
                System.out.println(j);
                break;
              }
            }
            break;
          }
          case 1 : {
            int index = sc.nextInt();
            int value = sc.nextInt();
            updateRMQ(1, n, 1, index, books[index - 1], value);
            books[index - 1] = value;
          }
        }
        
      }
    }
    sc.close();
  }
  
  static void printArray(short[] array) {
    for (int i = 0; i < array.length; i++) {
      System.out.print(i + ":" + array[i] + "\t");
    }
    System.out.println();
  }
  
  static void buildRMQ(int start, int end, int index) {
    aux[index] = new short[MAX_BOOK];
    if (start == end) {
      aux[index][books[start - 1]]++;
    } else {
      int mid = (start + end) >> 1;
      int left = index << 1;
      int right = left | 1;
      buildRMQ(start, mid, left);
      buildRMQ(mid + 1, end, right);
      for (int i = 0; i < MAX_BOOK; i++) {
        aux[index][i] = (short) (aux[left][i] + aux[right][i]);
      }
    }
  }
  
  static void updateRMQ(int start, int end, int index, int idx, int pvalue,
      int value) {
    if (start == end && start == idx) {
      aux[index][pvalue]--;
      aux[index][value]++;
    } else if (start <= idx && idx <= end) {
      int mid = (start + end) >> 1;
      int left = index << 1;
      int right = left | 1;
      updateRMQ(start, mid, left, idx, pvalue, value);
      updateRMQ(mid + 1, end, right, idx, pvalue, value);
      aux[index][pvalue]--;
      aux[index][value]++;
    }
  }
  
  static void queryRMQ(int start, int end, int index, int qstart, int qend,
      short[] result) {
    if (end < qstart || qend < start) {
    } else if (qstart <= start && end <= qend) {
      for (int i = 1; i < MAX_BOOK; i++) {
        result[i] += aux[index][i];
      }
    } else {
      int mid = (start + end) >> 1;
      int left = index << 1;
      int right = left | 1;
      queryRMQ(start, mid, left, qstart, qend, result);
      queryRMQ(mid + 1, end, right, qstart, qend, result);
    }
  }
}

File: scala/Library.scala
import java.util.Scanner
import scala.collection.immutable.List
import java.util.Collections
import java.util.{List => JList}

object Library{
  def main(args:Array[String]){
    val sc = new Scanner(System.in);
    for(tc <- 1 to sc.nextLine().toInt){
      val n = sc.nextLine().toInt;
      val books = sc.nextLine().split(" ").map(x=>Integer.parseInt(x));
      val aux = new Array[List[Int]](4*n);
      // build Our Range Minimum Queries
      buildRMQ(books, aux, 1, n, 1);
      for(nq <- 1 to sc.nextLine().toInt){
        val query = sc.nextLine().split(" ").map(x=>Integer.parseInt(x));
        query(0) match {
          case 0 => {
            System.out.println(getKth(queryRMQ(aux, 1, n, 1, query(1), query(2)), query(3)));
          }
          case 1 => {
            updateRMQ(aux, 1, n, 1, query(1), books(query(1)-1), query(2));
          }
        }
      }
    }
  }
  
  import _root_.scala.collection.JavaConversions._
  import java.util.{Collections, List => JList}
  class SearchableSeq[T](a: Seq[T])(implicit ordering: Ordering[T]) {
    val list: JList[T] = a.toList;
    def binarySearch(key: T): Int = Collections.binarySearch(list, key, ordering);
  }
  implicit def seqToSearchable[T](a: Seq[T])(implicit ordering: Ordering[T]) = new SearchableSeq(a)(ordering);
  
  /**
   * Return Kth rank element from sorted list 
   */
  def getKth(sortedList:Array[List[Int]], k:Int):Int={
  	var lists = Array[List[Int]]();
  	var maxLenIdx = 0; var maxLen = 0;
    for(i <- 0 to sortedList.length-1){
    	val list = sortedList(i).slice(0, k);
    	if(!list.isEmpty){
    		lists=lists++Array(list);
        if(maxLen<list.length){
          maxLen = list.length;
          maxLenIdx = lists.size-1;
        }
    	}
    }
    if(maxLen<=1){
      val finallist = lists.flatten.sorted;
      return finallist(k-1);
    }
    val middle = lists(maxLenIdx)(maxLen/2);
    var lowerHalfs = Array[List[Int]]();
    var upperHalfs = Array[List[Int]]();
    var lowerLen = 0;
    for(i <- 0 to lists.length-1){
      var insertIdx = lists(i).binarySearch(middle);
      if(insertIdx<0) insertIdx = -insertIdx - 1;
      val lower = lists(i).slice(0,insertIdx);
      val upper = lists(i).slice(insertIdx, lists(i).length);
      if(!lower.isEmpty){
      	lowerHalfs=lowerHalfs++Array(lower);
        lowerLen += lower.length;
      }
      if(!upper.isEmpty){
        upperHalfs=upperHalfs++Array(upper);
      }
    }
    
    if(lowerLen==k-1) return middle;
    else if(lowerLen>=k) return getKth(lowerHalfs, k);
    else return getKth(upperHalfs, k-lowerLen);
  }
  
  def buildRMQ(data:Array[Int], aux:Array[List[Int]], start:Int, end:Int, index:Int){
    if(start==end){
      aux(index) = List(data(start-1));
    }else{
      val mid = (start+end)>>1; val left = index<<1; val right = left|1;
      buildRMQ(data, aux, start, mid, left);
      buildRMQ(data, aux, mid+1, end, right);
      aux(index)=(aux(left):::aux(right)).sorted;
    }
  }
  
  def updateRMQ(aux:Array[List[Int]], start:Int, end:Int, index:Int, idx:Int, prevalue:Int, value:Int){
    if(start==end && start==idx){
      aux(index) = List(value);
    }else if(start<end){
      val mid = (start+end)>>1; val left = index<<1; val right = left|1;
      updateRMQ(aux, start, mid, left, idx, prevalue, value);
      updateRMQ(aux, mid+1, end, right, idx, prevalue, value);
      var insertIdx = aux(index).binarySearch(prevalue);
      aux(index) = aux(index).slice(0, insertIdx):::aux(index).slice(insertIdx+1, aux(index).length);
      insertIdx = aux(index).binarySearch(value);
      if(insertIdx==0){
        aux(index)= value +: aux(index);
      }else if(insertIdx == aux(index).length){
        aux(index)=aux(index) :+ value;
      }else{
      	if(insertIdx<0) insertIdx = -insertIdx - 1;
        aux(index)= (aux(index).slice(0,insertIdx):+value):::aux(index).slice(insertIdx, aux(index).length);
      }
    }
  }
  
  def queryRMQ(aux:Array[List[Int]], start:Int, end:Int, index:Int, qstart:Int, qend:Int):Array[List[Int]]={
      if(end<qstart || qend<start) Array(List[Int]());
      else if(qstart<=start && end<=qend) Array(aux(index));
      else{
        val mid = (start+end)>>1; val left = index<<1; val right = left|1;
        queryRMQ(aux, start, mid, left, qstart, qend)++queryRMQ(aux, mid+1, end, right, qstart, qend);
      }
  }
}